# 1 Análise Exploratória dos dados

A partir da base de dados precos_carros_brasil.csv, execute as seguintes tarefas:

In [1]:
# Importações
import pandas as pd # biblioteca de manipulação de dados
import warnings
warnings.filterwarnings('ignore') # para ignorar avisos de código e não poluir a saída do notebook

## a) Carregue a base

In [2]:
# Função para ler csv
dados = pd.read_csv('precos_carros_brasil.csv')
dados.shape # mostrar o número de linhas e colunas do dataframe

(267542, 11)

In [3]:
dados.head() # mostrar as primeiras linhas do dataframe

,year_of_reference,month_of_reference,fipe_code,authentication,brand,model,fuel,gear,engine_size,year_model,avg_price_brl
0,2021.0,January,004001-0,cfzlctzfwrcp,GM - Chevrolet,Corsa Wind 1.0 MPFI / EFI 2p,Gasoline,manual,1,2002.0,9162.0
1,2021.0,January,004001-0,cdqwxwpw3y2p,GM - Chevrolet,Corsa Wind 1.0 MPFI / EFI 2p,Gasoline,manual,1,2001.0,8832.0
2,2021.0,January,004001-0,cb1t3xwwj1xp,GM - Chevrolet,Corsa Wind 1.0 MPFI / EFI 2p,Gasoline,manual,1,2000.0,8388.0
3,2021.0,January,004001-0,cb9gct6j65r0,GM - Chevrolet,Corsa Wind 1.0 MPFI / EFI 2p,Alcohol,manual,1,2000.0,8453.0
4,2021.0,January,004003-7,g15wg0gbz1fx,GM - Chevrolet,Corsa Pick-Up GL/ Champ 1.6 MPFI / EFI,Gasoline,manual,"1,6",2001.0,12525.0


## b) Verifique se há valores faltantes nos dados. Caso haja, escolha uma tratativa para resolver o problema de valores faltantes

In [4]:
# Verificando valores faltantes nos dados
dados.isna().any()

year_of_reference     True
month_of_reference    True
fipe_code             True
authentication        True
brand                 True
model                 True
fuel                  True
gear                  True
engine_size           True
year_model            True
avg_price_brl         True
dtype: bool

In [5]:
# Todas as colunas possuem valores faltantes, vamos verificar a quantidade de valores faltantes em cada coluna
dados.isna().sum()

year_of_reference     65245
month_of_reference    65245
fipe_code             65245
authentication        65245
brand                 65245
model                 65245
fuel                  65245
gear                  65245
engine_size           65245
year_model            65245
avg_price_brl         65245
dtype: int64

In [6]:
# Número de linhas com todos os valores faltantes
dados[dados.isna().all(axis=1)].shape

(65245, 11)

Constatei que os dados faltantes são várias linhas que possuem TODAS as colunas faltantes. (Número de linhas com todos os valores faltantes é igual ao número de valores faltantes em todas as colunas)
Portanto, faz mais sentido removê-las do que tentar inferir algum dado.

In [7]:
dados=dados.dropna() # remover linhas com valores faltantes
dados.isna().sum() # comprova que o número de valores faltantes agora é zero

year_of_reference     0
month_of_reference    0
fipe_code             0
authentication        0
brand                 0
model                 0
fuel                  0
gear                  0
engine_size           0
year_model            0
avg_price_brl         0
dtype: int64

In [8]:
# Conferindo o número de linhas e colunas do dataframe após a limpeza
dados.shape

(202297, 11)

## c) Verifique se há dados duplicados nos dados

In [ ]:
# Soma a quantidade de dados duplicados (linhas que possuem os mesmos valores em todas as colunas)
dados.duplicated().sum()

np.int64(2)

In [10]:
# Como são apenas 2 linhas, quero dar uma olhada nelas para decidir se quero removê-las ou não
dados[dados.duplicated()].head()

,year_of_reference,month_of_reference,fipe_code,authentication,brand,model,fuel,gear,engine_size,year_model,avg_price_brl
45793,2021.0,June,025232-8,5rtdwkpkpq5h,Renault,DUSTER OROCH Dyna. 2.0 Flex 16V Mec.,Gasoline,manual,2,2018.0,69893.0
189896,2022.0,December,003296-4,3r6c277cnqcb,Ford,Ranger Limited 3.0 PSE 4x4 CD TB Diesel,Diesel,manual,3,2007.0,64638.0


In [11]:
# OK... vamos removê-las!
dados.drop_duplicates(inplace=True)
# Obs: O parâmetro inplace=True faz com que a remoção seja feita no próprio dataframe, sem precisar atribuir o resultado a uma nova variável

In [12]:
# Conferindo o número de linhas e colunas do dataframe após a limpeza
dados.shape

(202295, 11)

No caso, reduziram de 202297 linhas para 202295, confirmando nossa remoção.

## d) Crie duas categorias, para separar colunas numéricas e categóricas. Imprima o resumo de informações das variáveis numéricas e categóricas (estatística descritiva dos dados)

In [13]:
# Vou dar uma olhadinha na tipagem dos dados antes de quebrar o dataset
dados.dtypes

year_of_reference     float64
month_of_reference        str
fipe_code                 str
authentication            str
brand                     str
model                     str
fuel                      str
gear                      str
engine_size               str
year_model            float64
avg_price_brl         float64
dtype: object

In [18]:
# O engine_size está como string, mas aparentemente é numérico. Vou verificar os valores únicos dessa coluna para entender o que está acontecendo
dados['engine_size'].unique()

array([1. , 1.6, 2.2, 4.3, 2.5, 1.8, 2. , 4.2, 3.8, 4.1, 5.7, 2.8, 2.4,
       1.4, 3.6, 6.2, 3. , 1.2, 1.5, 1.3, 1.9, 2.3, 4. , 3.9, 5. , 3.5,
       3.2, 2.7, 3.3])

In [ ]:
# O engine_size tem valores numéricos, o que pode ser uma boa variável para o modelo (ex: quanto maior o motor, mais caro o carro).
# Vou converter o separador decimal de vírgula para ponto e depois converter a coluna para numérica
dados['engine_size'] = dados['engine_size'].str.replace(',', '.').astype(float)

AttributeError: Can only use .str accessor with string values, not floating

In [17]:
dados.dtypes # conferindo a tipagem dos dados após a conversão

year_of_reference     float64
month_of_reference        str
fipe_code                 str
authentication            str
brand                     str
model                     str
fuel                      str
gear                      str
engine_size           float64
year_model            float64
avg_price_brl         float64
dtype: object

In [22]:
# Outro dado que também precisa de conversão é o month_of_reference que está em string.
# Vou criar uma coluna nova chamada month_of_reference_num com os meses convertidos para números (ex: janeiro = 1, fevereiro = 2, etc)
# Assim a coluna pode ser usada para o modelo, trazendo uma ordem de grandeza para os meses (ex: quanto mais avançado no ano, menor o valor do carro).

# Criando um dicionário para mapear os meses em números
meses = {
    'January': 1,
    'February': 2,
    'March': 3,
    'April': 4,
    'May': 5,
    'June': 6,
    'July': 7,
    'August': 8,
    'September': 9,
    'October': 10,
    'November': 11,
    'December': 12
}
# Criando a nova coluna month_of_reference_num usando o dicionário de meses
dados['month_of_reference_num'] = dados['month_of_reference'].map(meses)


In [ ]:
# Conferindo se deu certo:
dados['month_of_reference_num'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [25]:
dados.dtypes # conferindo a tipagem dos dados após a conversão

year_of_reference         float64
month_of_reference            str
fipe_code                     str
authentication                str
brand                         str
model                         str
fuel                          str
gear                          str
engine_size               float64
year_model                float64
avg_price_brl             float64
month_of_reference_num      int64
dtype: object

In [26]:
# Criando categorias para separar numéricas e categóricas
numericas_cols = [col for col in dados.columns if dados[col].dtype != 'str']
categoricas_cols = [col for col in dados.columns if dados[col].dtype == 'str']

In [27]:
numericas_cols

['year_of_reference',
 'engine_size',
 'year_model',
 'avg_price_brl',
 'month_of_reference_num']

In [28]:
categoricas_cols

['month_of_reference',
 'fipe_code',
 'authentication',
 'brand',
 'model',
 'fuel',
 'gear']

## e) Imprima a contagem de valores por modelo (model) e marca do carro (brand)

In [ ]:
# Exbindo a contagem de valores por modelo:
dados['model'].value_counts()

model
Palio Week. Adv/Adv TRYON 1.8 mpi Flex    425
Focus 1.6 S/SE/SE Plus Flex 8V/16V 5p     425
Focus 2.0 16V/SE/SE Plus Flex 5p Aut.     400
Saveiro 1.6 Mi/ 1.6 Mi Total Flex 8V      400
Corvette 5.7/ 6.0, 6.2 Targa/Stingray     375
                                         ... 
STEPWAY Zen Flex 1.0 12V Mec.               2
Saveiro Robust 1.6 Total Flex 16V CD        2
Saveiro Robust 1.6 Total Flex 16V           2
Gol Last Edition 1.0 Flex 12V 5p            2
Polo Track 1.0 Flex 12V 5p                  2
Name: count, Length: 2112, dtype: int64

In [ ]:
# Exbindo a contagem de valores por marca:
dados['brand'].value_counts()

brand
Fiat               44962
VW - VolksWagen    44312
GM - Chevrolet     38590
Ford               33150
Renault            29191
Nissan             12090
Name: count, dtype: int64

## f) Dê um breve explicação (máximo de quatro linhas) sobre os principais resultados encontrados na Análise Exploratória dos dados

Foi constatada e corrigida a presença de registros faltantes no dataset. Após verificarmos se tratar de linhas inteiras vazias, tomamos a decisão de excluí-las junto com 2 linhas que estavam duplicadas. Constatamos que engine_size e month_of_reference estavam sendo tratadas como string e foram convertidas para tratamento como numérico.

### Considerações completas
A primeira constatação foi a presença de registros faltantes no dataset. Ao verificar, constatamos que se tratavam de linhas inteiras, com todas as colunas vazias, e não simples valores faltantes em determinadas colunas.
Essa descoberta nos levou a tomar a decisão de excluí-las ao invés de tentar alguma abordagem de preenchimento (como Moda, Média, etc).
Também detectou-se 2 linhas duplicadas nos dados que foram prontamente removidas.
Após isso, constatamos que o engine_size estava tipado como string. Sendo todos os seus valores numéricos, convertemos para float (substituindo o separador). Isso permite que a engine_size seja utilizada como uma variável de grandeza no modelo ao invés de categórica.
Por fim, a month_of_reference também estavam como texto, o que poderia ser interpretado como uma simples categoria e se perder a ordem crescente dos meses, dificultando na visualização de dados, por exemplo. Foi criada uma coluna auxiliar month_of_reference_num.

### Considerações para visualização
A nova coluna month_of_reference_num pode ser usada para criar visualizações por meses, de forma ordenada.

### Considerações para treinamento de modelo
Dataset limpo é o dataframe dados
Colunas catégoricas listadas em: categorical_columns
Colunas numéricas listadas em: numerical_columns
Atenção que há uma coluna month_of_reference e outra month_of_reference_num com a mesma informação, apenas com tipos de dados diferentes. Necessário decidir qual vai ser usada para o treinamento.